File: Strings/1021. Remove Outermost Parentheses/1021. Remove Outermost Parentheses.py
class Solution:
    def removeOuterParentheses(self, s: str) -> str:
        ans = []
        depth = 0

        for char in s:
            if char == "(":
                if depth > 0:
                    ans.append(char)
                depth += 1
            else:
                depth -= 1
                if depth > 0:
                    ans.append(char)

        return "".join(ans)

# Time: O(n)
# Space: O(n)

File: Strings/1021. Remove Outermost Parentheses/1021. Remove Outermost Parentheses.js
/**
 * @param {string} s
 * @return {string}
 */
var removeOuterParentheses = function (s) {
  const ans = [];
  let depth = 0;

  for (const char of s) {
    if (char === "(") {
      if (depth > 0) ans.push(char);
      depth++;
    } else {
      depth--;
      if (depth > 0) ans.push(char);
    }
  }

  return ans.join("");
};

// Time: O(n)
// Space: O(n)
